src/jit/optimizer.cpp

```diff
@@ -1574,179 +1588,165 @@ void Compiler::optFindNaturalLoops()
+                    */
+                    if (block->bbNum > oldBlockMaxNum)
                     {
-                        canIterateLoop = true;
-                        break;
+                        // This is a new block we added to connect fall-through, so the
+                        // recorded dominator information doesn't cover it.  Just continue,
+                        // and when we process its unique predecessor we'll abort if ENTRY
+                        // doesn't dominate that.
                     }
-                    else if (predEntry->flBlock != head)
+                    else if (!fgDominate(entry, block))
                     {
-                        // The entry block has multiple predecessors outside the loop; the 'head'
-                        // block isn't the only one. We only support a single 'head', so bail.
                         goto NO_LOOP;
                     }
-                }
-
-                if (!canIterateLoop)
-                {
-                    goto NO_LOOP;
-                }
-
-                /* Double check - make sure that all loop blocks except ENTRY
-                 * have no predecessors outside the loop - this ensures only one loop entry and prevents
-                 * us from considering non-loops due to incorrectly assuming that we had a back edge
-                 *
-                 * OBSERVATION:
-                 *    Loops of the form "while (a || b)" will be treated as 2 nested loops (with the same header)
-                 */
 
-                for (loopBlock = top; loopBlock != bottom->bbNext; loopBlock = loopBlock->bbNext)
-                {
-                    if (loopBlock == entry)
+                    // Add preds to the worklist, checking for side-entries.
+                    for (flowList* predIter = block->bbPreds; predIter != nullptr; predIter = predIter->flNext)
                     {
-                        continue;
-                    }
+                        BasicBlock* pred = predIter->flBlock;
 
-                    for (predTop = loopBlock->bbPreds; predTop != nullptr; predTop = predTop->flNext)
-                    {
-                        if (predTop->flBlock->bbNum < top->bbNum || predTop->flBlock->bbNum > bottom->bbNum)
+                        unsigned int testNum = positionNum(pred);
+
+                        if ((testNum < top->bbNum) || (testNum > bottom->bbNum))
                         {
-                            // noway_assert(!"Found loop with multiple entries");
+                            // Pred is out of loop range
+                            if ((pred == head) && (block == entry))
+                            {
+                                // This is the single entry we expect.
+                                continue;
+                            }
+
+                            // There are multiple entries to this loop, don't consider it.
```

Comment:
This "side entry" could also be from a disjoint part of the loop. It might be interesting someday to work through relaxing this condition. Something like: let the pred closure happen, since it seems like the new dominates check should prevent wandering outside the loop, and see how many disjoint bodies there are out there.

If these are common, we may now have the machinery to compact them too.